### DISCUSSION TITLE
後端無法取得 cookie

### DISCUSSION THREAD

**Discussion by EshauHuang** [Lidemy/forum#221] [後端相關問題] [answered]
打擾各位了，遇到一些 cookie 的問題想要請教。

目前使用 jwt 在使用者登入時會自動將 accessToken 帶在 cookie

```js
const setAccessToken = (req, res, accessToken) => {
  return res.cookie("accessToken", accessToken, {
    httpOnly: true,
  });
};
```

當我使用 postmen 發 request 登入時，會確實新增一個帶著 accessToken 的 cookie ，之後再次發 request 至後端來取得使用者資料時，前端可以正確接到使用者的資料。但當我用網頁的 api 試著以 fetch 來操作登入及取得使用者時，發現登入可以正常運行，但取得使用者資料會是回傳自己設定的錯誤訊息，原因是後端接收不到 cookie(試著印出 `req.cookie` 會顯示`[Object: null prototype] {}`)。

從 network 來看接收到的資料，發現 signin 回傳的資料確實帶著 cookie，但 me 卻沒帶著 cookie 發送資料，而我也有去看了 application -> Storage -> Cookies 沒有發現任何的 cookie 資料，不知道是不是沒有正確加進去。

### signin

![Imgur](https://i.imgur.com/bSfG1Xg.png)

![Imgur](https://i.imgur.com/POsDpet.png)

### me

![Imgur](https://i.imgur.com/yWOMh1N.png)

這是抓使用者資料的 api

```js
const response = await fetch(`http://localhost:3333/me`, {
      method: "GET",
      credentials: "include",
      headers: {
        "Content-Type": "application/json",
      },
    });
```

後端的 cors 設定

```
server.use(
  cors({
    origin: "*",
    credentials: true,
  })
);
```

自己在猜應該是後端帶入 cookie 時可能少帶了什麼東西，所以瀏覽器本該幫我加上 cookie 的，但因為某種原因所以沒有帶上，另外我的 postmen 有下載一個 Postman Agent 不知道是不是因為這樣 postmen 才能正確執行，麻煩各位的幫助了🙏，非常感謝。

**Comment by aszx87410** [comment, marked as answer]
console 沒有出現錯誤訊息嗎？
可參考：https://blog.huli.tw/2021/02/19/cors-guide-3/ 裡面 Day3：帶上 Cookie 的部分

> 錯誤訊息其實已經解釋得很清楚了，如果要帶上 cookie 的話，那 Access-Control-Allow-Origin 不能是 *，一定要明確指定 origin。

**Reply by aszx87410** [reply]
不對，我現在看到雖然後端是設定 * 不過 response header 是有指定 origin 的

**Reply by EshauHuang** [reply]
感謝老師，「我現在看到雖然後端是設定 * 不過 response header 是有指定 origin 的」這部分是我截錯圖了，因為我 `origin: '*'` 與 `origin: 'http://localhost:3000'` 都設置過了但都沒有用。

log 沒有錯訊息，但會顯示我設置的錯誤訊息 `UNAUTHORIZED`，代表應該是有正常進到後端去
```js
const isAuth = (req) => {
  const { accessToken } = req.cookies;
  if (!accessToken) throw new Error("UNAUTHORIZED");
  const { iat, exp, ...user } = verify(
    accessToken,
    process.env.ACCESS_TOKEN_SECRET
  );
  return user;
};
```

**Reply by aszx87410** [reply]
無論你 origin 是 * 還是 localhost:3000，一定都會正常進到後端去，這其實不影響你問的這個問題
但如果是設成 * 的話 console 應該會有錯誤訊息出現
然後以後提問要記得 code 跟圖片要一致，不然變因會太多

**Reply by EshauHuang** [reply]
瞭解了，會再加強問問題的技巧。

**Comment by aszx87410** [comment]
> 而我也有去看了 application -> Storage -> Cookies 沒有發現任何的 cookie 資料，不知道是不是沒有正確加進去。
這邊你第二張截圖看起來 cookie 是有存進去的，跟敘述好像不太一樣？

另外，你沒有設定過期時間，所以過期時間會由瀏覽器決定，這個可以設定一下

**Reply by aszx87410** [reply]
最後撇除這些，也有可能是 same site 的問題，可以換個瀏覽器試試看

**Reply by EshauHuang** [reply]
原來 cookie 另外也有過期時間，我還以為 jwt 設定了就沒問題了，感謝老師。

**Comment by EshauHuang** [comment]
感謝老師，目前測試下來可以接受到 cookie 了，而發現了一個小地方不太懂，因為我登入的流程是：

前端發 request 到 /signin -> -> 後端驗證使用者 -> 後端加上 cookie -> 前端發 request 到 /me -> 後端驗證 jwt -> 後端回傳驗證後訊息 -> 前端接收到使用者資料

而一開始我只在 me 這部分加上 `res.header("Access-Control-Allow-Credentials", true);` 並修正 origin 的 '*'，但發現還是與之前一樣「能接收到 cookie 但瀏覽器內的 cookie 沒有東西」。在此其間我修了很多東西，將 res.cookie 加上 maxAge、expire、sameSite、secure，其中比較奇怪的地方是當我加上 sameSite、secure 時，我的 postmen 反而拿不到 cookie，不過前面的參數還是沒有解決沒有 cookie 的問題，最後我試著在 sign 這條 api 也加上 `credentials: "include"`，而後端也做相應變化後，就可以存取到 cookie。

而我的問題是為何 signin 這條 api 需要加入 `credentials: "include"`，明明我沒有要帶上 cookie 請求，而是要接收 cookie，另外我以為瀏覽器會不會儲存 cookie 的原因是 `res.header("Access-Control-Allow-Credentials", true);`，所以我把 `credentials: "include"` 刪掉，只在後端加上 `Access-Control-Allow-Credentials`，但這樣還是沒成功，最後變成 signin 與 me 這兩條 api 都必須加上這些參數，感謝老師的指導。

**Reply by aszx87410** [reply]
https://medium.com/d-d-mag/%E5%92%8C-cors-%E8%B7%9F-cookie-%E6%89%93%E4%BA%A4%E9%81%93-dd420ccc7399

> 不過如果你在 a 網域送出了 b 網域的請求，且 b 網域回傳了 cookie 的訊息，那麼在 a 網域會以 b 網域的形式儲存一份 cookie，如果沒有設定 withCredentials 或是 credentials: ‘include’ 的話，就算伺服器有回傳 Set-Cookie，一樣不會被寫入

**Reply by EshauHuang** [reply]
感謝老師🙏
